src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsaccreditedprogrammesapi/domain/CourseService.kt

```diff
@@ -54,7 +54,18 @@ class CourseService(
     courses.forEach { it.prerequisites.clear() }
   }
 
-  fun replaceAllOfferings(offeringRecords: List<OfferingRecord>) {
-    offeringRecords.size
+  fun replaceAllOfferings(replacements: List<OfferingRecord>) {
+    val allCourses = courseRepository.allCourses()
+    clearOfferings(allCourses)
+    val coursesByName = allCourses.associateBy(CourseEntity::name)
+    replacements.forEach { record ->
+      coursesByName[record.course]?.run {
+        offerings.add(Offering(organisationId = record.prisonId, contactEmail = record.contactEmail ?: ""))
```

Comment:
As mentioned earlier, `contactEmail` should always be present. I know it's not in the current CSV, so how about we log it in the same way we doing for prerequisites for which a matching course can't be found?

Reply:
We could.  The presence or lack of an e-mail address is very obvious to anyone working with the spreadsheets and it is easy to remedy.  Do we really need extra checking here? If so then let's do this as a separate story/card.

Reply:
I suppose, but if we're not worried about the presence of lack of an email address, do we need a default empty string value for it?